File: TS/系统/03．技能系统/00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/15．单位运行时上下文工厂.ts
/** @noSelfInFile */

import { 设置单位技能壳普通提示 } from "../../02．通用函数/15．单位技能壳提示";
import { 创建机制清理篮子, type 机制清理篮子 } from "../06．机制清理/01．机制清理篮子";

const jass = require("jass.common") as any;

const GetHandleId = jass.GetHandleId as (whichHandle: any) => number;
const { registerDeathListener } = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心") as {
  registerDeathListener: (this: void, callback: (this: void, dyingUnit: any, killingUnit: any) => void) => void;
};

export interface 单位运行时上下文基础 {
  清理: 机制清理篮子;
}

export interface 单位运行时上下文工厂参数<T extends 单位运行时上下文基础> {
  名称: string;
  主动技能提示?: any;
  创建上下文: (this: void, unit: any, 清理: 机制清理篮子) => T;
  on创建?: (this: void, context: T) => void;
  on清理?: (this: void, context: T) => void;
  /**
   * 上下文对应单位死亡时触发；只会为已经登记到本工厂的单位调用。
   * 是否自动清理由“死亡时自动清理”单独控制，方便保留死亡后的延迟收束。
   */
  on单位死亡?: (this: void, context: T, dyingUnit: any, killingUnit: any) => void;
  死亡时自动清理?: boolean;
}

export interface 单位运行时上下文工厂<T extends 单位运行时上下文基础> {
  获取(this: void, unit: any): T | undefined;
  获取或创建(this: void, unit: any): T | undefined;
  获取全部(this: void): T[];
  清理上下文(this: void, unit: any): void;
  取单位ID(this: void, unit: any): number;
}

function 默认取单位ID(this: void, unit: any): number {
  if (unit == null || unit === 0) return 0;
  return GetHandleId(unit) || 0;
}

export function 创建单位运行时上下文工厂<T extends 单位运行时上下文基础>(
  this: void,
  参数: 单位运行时上下文工厂参数<T>,
): 单位运行时上下文工厂<T> {
  const 上下文表: Record<number, T | undefined> = {};

  function 获取(this: void, unit: any): T | undefined {
    const id = 默认取单位ID(unit);
    return id === 0 ? undefined : 上下文表[id];
  }

  function 获取或创建(this: void, unit: any): T | undefined {
    const id = 默认取单位ID(unit);
    if (id === 0) return undefined;
    let context = 上下文表[id];
    if (context != null) return context;
    const 清理 = 创建机制清理篮子(参数.名称);
    context = 参数.创建上下文(unit, 清理);
    if (参数.主动技能提示 != null) 设置单位技能壳普通提示(unit, 参数.主动技能提示);
    上下文表[id] = context;
    if (参数.on创建 != null) 参数.on创建(context);
    return context;
  }

  function 获取全部(this: void): T[] {
    const result: T[] = [];
    for (const key in 上下文表) {
      const context = 上下文表[key];
      if (context != null) result.push(context);
    }
    return result;
  }

  function 清理上下文(this: void, unit: any): void {
    const id = 默认取单位ID(unit);
    if (id === 0) return;
    const context = 上下文表[id];
    if (context != null) {
      if (参数.on清理 != null) 参数.on清理(context);
      context.清理.清理全部();
    }
    delete 上下文表[id];
  }

  function 处理单位死亡(this: void, dyingUnit: any, killingUnit: any): void {
    const id = 默认取单位ID(dyingUnit);
    if (id === 0) return;
    const context = 上下文表[id];
    if (context == null) return;
    if (参数.on单位死亡 != null) 参数.on单位死亡(context, dyingUnit, killingUnit);
    if (参数.死亡时自动清理 && 上下文表[id] === context) 清理上下文(dyingUnit);
  }

  if (参数.死亡时自动清理 || 参数.on单位死亡 != null) {
    registerDeathListener(处理单位死亡);
  }

  return {
    获取,
    获取或创建,
    获取全部,
    清理上下文,
    取单位ID: 默认取单位ID,
  };
}

export {};

File: src/系统/03．技能系统/00．技能模板+函数/04．机制组件/10．复杂战斗通用机制/15．单位运行时上下文工厂.lua
local ____lualib = require("lualib_bundle")
local __TS__Delete = ____lualib.__TS__Delete
local ____exports = {}
local ____15_FF0E_5355_4F4D_6280_80FD_58F3_63D0_793A = require("系统.03．技能系统.00．技能模板+函数.02．通用函数.15．单位技能壳提示")
local _____8BBE_7F6E_5355_4F4D_6280_80FD_58F3_666E_901A_63D0_793A = ____15_FF0E_5355_4F4D_6280_80FD_58F3_63D0_793A["设置单位技能壳普通提示"]
local ____01_FF0E_673A_5236_6E05_7406_7BEE_5B50 = require("系统.03．技能系统.00．技能模板+函数.04．机制组件.06．机制清理.01．机制清理篮子")
local _____521B_5EFA_673A_5236_6E05_7406_7BEE_5B50 = ____01_FF0E_673A_5236_6E05_7406_7BEE_5B50["创建机制清理篮子"]
local jass = require("jass.common")
local GetHandleId = jass.GetHandleId
local ____require_result_0 = require("系统.00．核心系统.01．事件中心.07．单位死亡事件中心")
local registerDeathListener = ____require_result_0.registerDeathListener
local function _____9ED8_8BA4_53D6_5355_4F4DID(unit)
    if unit == nil or unit == 0 then
        return 0
    end
    return GetHandleId(unit) or 0
end
____exports["创建单位运行时上下文工厂"] = function(_____53C2_6570)
    local _____4E0A_4E0B_6587_8868 = {}
    local function _____83B7_53D6(unit)
        local id = _____9ED8_8BA4_53D6_5355_4F4DID(unit)
        local ____temp_1
        if id == 0 then
            ____temp_1 = nil
        else
            ____temp_1 = _____4E0A_4E0B_6587_8868[id]
        end
        return ____temp_1
    end
    local function _____83B7_53D6_6216_521B_5EFA(unit)
        local id = _____9ED8_8BA4_53D6_5355_4F4DID(unit)
        if id == 0 then
            return nil
        end
        local context = _____4E0A_4E0B_6587_8868[id]
        if context ~= nil then
            return context
        end
        local _____6E05_7406 = _____521B_5EFA_673A_5236_6E05_7406_7BEE_5B50(_____53C2_6570["名称"])
        context = _____53C2_6570["创建上下文"](unit, _____6E05_7406)
        if _____53C2_6570["主动技能提示"] ~= nil then
            _____8BBE_7F6E_5355_4F4D_6280_80FD_58F3_666E_901A_63D0_793A(unit, _____53C2_6570["主动技能提示"])
        end
        _____4E0A_4E0B_6587_8868[id] = context
        if _____53C2_6570["on创建"] ~= nil then
            _____53C2_6570["on创建"](context)
        end
        return context
    end
    local function _____83B7_53D6_5168_90E8()
        local result = {}
        for key in pairs(_____4E0A_4E0B_6587_8868) do
            local context = _____4E0A_4E0B_6587_8868[key]
            if context ~= nil then
                result[#result + 1] = context
            end
        end
        return result
    end
    local function _____6E05_7406_4E0A_4E0B_6587(unit)
        local id = _____9ED8_8BA4_53D6_5355_4F4DID(unit)
        if id == 0 then
            return
        end
        local context = _____4E0A_4E0B_6587_8868[id]
        if context ~= nil then
            if _____53C2_6570["on清理"] ~= nil then
                _____53C2_6570["on清理"](context)
            end
            local ____self_2 = context["清理"]
            ____self_2["清理全部"](____self_2)
        end
        __TS__Delete(_____4E0A_4E0B_6587_8868, id)
    end
    local function _____5904_7406_5355_4F4D_6B7B_4EA1(dyingUnit, killingUnit)
        local id = _____9ED8_8BA4_53D6_5355_4F4DID(dyingUnit)
        if id == 0 then
            return
        end
        local context = _____4E0A_4E0B_6587_8868[id]
        if context == nil then
            return
        end
        if _____53C2_6570["on单位死亡"] ~= nil then
            _____53C2_6570["on单位死亡"](context, dyingUnit, killingUnit)
        end
        if _____53C2_6570["死亡时自动清理"] and _____4E0A_4E0B_6587_8868[id] == context then
            _____6E05_7406_4E0A_4E0B_6587(dyingUnit)
        end
    end
    if _____53C2_6570["死亡时自动清理"] or _____53C2_6570["on单位死亡"] ~= nil then
        registerDeathListener(_____5904_7406_5355_4F4D_6B7B_4EA1)
    end
    return {
        ["获取"] = _____83B7_53D6,
        ["获取或创建"] = _____83B7_53D6_6216_521B_5EFA,
        ["获取全部"] = _____83B7_53D6_5168_90E8,
        ["清理上下文"] = _____6E05_7406_4E0A_4E0B_6587,
        ["取单位ID"] = _____9ED8_8BA4_53D6_5355_4F4DID
    }
end
return ____exports
